h5io_browser/base.py

```diff
@@ -229,6 +230,44 @@ def _merge_nested_dict(main_dict: dict, add_dict: dict) -> dict:
             main_dict[k] = v
     return main_dict
 
+from collections import defaultdict
+_FILE_HANDLE_CACHE_COUNTER = defaultdict(int)
+_FILE_HANDLE_CACHE = {}
+import threading
+_FILE_HANDLE_CACHE_LOCK = threading.RLock()
+
+import contextlib
+@contextlib.contextmanager
+def WithOpenHDF(filename: str, mode: str = "r", swmr: bool = False):
+    filename = os.path.realpath(os.path.abspath(filename))
+    with _FILE_HANDLE_CACHE_LOCK:
+        file_is_open = filename in _FILE_HANDLE_CACHE
+    # File already cached, someone above us opened it and will clean it up, we just need to pass it down
+    if file_is_open:
+        # print("EVERYBODY HIT THE CACHE")
+        with _FILE_HANDLE_CACHE_LOCK:
+            handle = _FILE_HANDLE_CACHE[filename]
```

Comment:
This is a classic mistake in multithreading. You can have a race condition between the check and the retrieval. You must do the `filename in _FILE_HANDLE_CACHE` and `handle =  _FILE_HANDLE_CACHE[key]` within the same locking.

Reply:
Naturally. :|